### DISCUSSION TITLE
WinUI Community Call September 2023 Q&A Follow-up

### DISCUSSION THREAD

**Discussion by duncanmacmichael** [microsoft/microsoft-ui-xaml#8980] [General]
Hi everyone, since we received a lot of questions on the September 2023 community call discussion post and in the call's live comments that we didn't have time to answer, we'd like to follow up here and make sure we address all the ones we couldn't get to during the call. We appreciate your patience and understanding as we go through a transition time with the call's format and cadence, and want to thank you as always for your great feedback and questions. 

> NOTE: If there is a question you asked but don't see here, we either can't answer it, it was a duplicate of another question, or it was already answered on the call itself.

## Questions from the September 2023 community call discussion post

These questions were asked on the [community call discussion post](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884) but we'll consolidate them here along with the other questions in an effort to make the answers clearer and not get lost in individual comment chains.

### Q: Are there any improvements to Xaml coming soon, It's been years since Xaml saw substantial improvement in WinUI/UWP Space.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7040015)

**A:** It seems like you're asking about XAML language features here from what we can tell. We have been making runtime updates, although there haven't been many updates to the language itself as you observed. We don't have any immediate plans here but this question has brought this to top of mind for us, thank you.

### Q: Yes, what they should be making up for now is missing uwp-equivalent features and performance fixes like maps, inkcanvas, validation, native aot support, etc.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7053417)

**A:** Thank you for bringing these topics up - we actually mentioned three out of four of them in the [roadmap for 1.5](https://github.com/microsoft/WindowsAppSDK/blob/main/docs/roadmap.md)! A maps control and inking are under investigation for 1.5, and we are working on native AOT support in that time frame as well. As for validation, we assume you mean text input for validation, which we don't have plans for currently.

### Q: Data Grid in WinUI please, but not implemented as a grid but like Apple CollectionViews with it's own separated, configurable layouts allowing to do circular collections, piles of items etc.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7053711)

**A:** We are looking into the table view control for 1.5, which is a lighter weight version of the data grid control.

### Q: It would be great if the team would address how they plan to improve how this project is handled...there is still no WinUI 3 source available...is the team planning to start actively working on the backlog? It feels like the development of WinUI has pretty much stopped.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7042782)

**A:** We touched base on many of these topics on the call itself but can recap them here. 

- For the source, we do have that coming very soon! It's going through the final internal hoops for sign-off and completion as of the time of writing this.
- The GitHub backlog cleanup process is ongoing, with PM and engineering working together on separating out the feature requests and bugs from the issues list. 
- Thank you for your thoughts on the direction of WinUI - we try to be as transparent as possible with roadmaps, answering questions, and fixing bugs, but we realize we can't tackle everything. The recent 1.5 roadmap and even this discussion post are examples of how we're striving to improve here and make sure we communicate how we're paying attention to the community. For long-standing bugs, we have begun mirroring issues from GitHub to our internal bug tracking system to allow for transparency on bugs' progress, and have even gotten to work on some old bugs thanks to our improved processes (such as a fix for localized accelerator keys coming in 1.4.2 - this was a very old bug!). Some of this sentiment might derive from periods where we're working on larger underlying structural issues that will allow us to improve things in the long-term, which means a period of low activity on some fronts, and we realize how that might make us appear unresponsive or quiet during those times. When those labors begin bearing fruit in the 1.4 to 1.5 time frame via new features, closing of feature gaps, and increased ability to fix bugs, we're hoping this makes our direction clearer and we thank you for your understanding.
- Although we see this is a very important point for you, unfortunately we can't comment on staffing questions. :)

### Q: I am extremely curious about the inking experience and why it isn't on the roadmap. I would think this would be an important addition so that their own first party apps such as OneNote can start taking advantage of WinUI3.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7049357)

**A:** Although ChaimCL partially addressed this in a reply to this question, we just wanted to clarify that we are looking into the inking experience in 1.5. Stay tuned!

### Q: "When can we publish the app as a PublishSingleFile? I have been waiting since version 1.0 was released!

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7049318)

**A:** We're working on this now and hope to have a fix included in an upcoming servicing release. Stay tuned and thanks for your patience!

### Q: It is more than 5 months, since Windows App SDK version 1.3 has been released and since that day we are reporting, that PasswordBox is not working properly for touch displays. ...

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7051723)

**A:** We actually fixed this as of 10/04/23! This did result in a different but related issue being opened up, per [this comment](https://github.com/microsoft/microsoft-ui-xaml/issues/6291#issuecomment-1747823639). We are actively investigating that new bug and have a potential fix in the works.

### Q: Please take a look at bug report #8832. Seems the Resource mechanism has some serious bugs. Apart from bug 8832, there are plenty of Resourse related issues, which take up lots of valuable time to find not-so-elegant work-arounds.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7053547)

**A:** This is on our radar, thank you for bringing it up. The work necessary to share the WinUI 3 source code resulted in a delay in some bugs, including this one. However, it is in our backlog and we don't have an exact time frame to commit to a fix at the moment. Thanks for bearing with us; we're aware of this bug.

### Q: The position of WinUI on the development landscape is still very uncertain in regard to Microsoft's approach to the new versions of their major products. ... We need reassurance. A good one. Not vague and lofty comments, but good proof of the company's vision.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7054943)

**A:** We understand your viewpoint here and are actively working on a plan to address these points in the coming months, collaborating on multiple levels internally including engaging with leadership. We want to make sure WinUI's position within the greater UX framework landscape is clear and make recommendations appropriately, but we need a little more time. This is part of our greater effort to clarify things with the community and is at the top of our list. Thanks for your astute observations and dedication to helping us streamline our direction.

### Q: We [developer name] have invested heavily in WinUI/WindowsAppSDK, and I get fewer and fewer stability related issues from our customers. But still, I keep seeing random XAML-related unhandled exceptions ... I have no clue on where in the XAML tree this happens and what it triggers. ... I'd rather see these random XAML-related issues disappear. :)

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7055892)

**A:** This is one of the primary reasons we're sharing source, for these types of crash situations! Debuggability is a major priority for us and we hope it helps to solve more of these difficult-to-diagnose exceptions.

### Q: Will we ever get a BlazorWebView for WinUI 3?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7065613)

**A:** We don't currently have any plans for direct BlazorWebView support beyond the MAUI support listed in the linked thread.

### Q: Any work planned on win2d? Looks like it is abandoned currently. It is currently the only managed interface to DWriteCore without building C++ wrappers.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7080593)

**A:** No, we don't have any feature work planned for Win2D at the moment; we largely completed the port to support WinUI 3 earlier this year (apart from a few controls that were not ported) and that was the extent of our plans for the time being.

### Q: Question about the new ScroIlView control. Is it possible to use the new ScrollView with SwapChainPanel to implement scrollable experience with animations (overpan/inertia scrolling etc.) over Direct3D content?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7094702)

**A:** One of our engineers responded to this directly on the link above!

### Q: I'm using WinUI 3 to build an app that uses multiple different user controls that utilize Direct3D to draw content. ... Unfortunately, bug #7924 was introduced after release 1.1 and to this day it prevents me from upgrading to any release after 1.1. ... Would it be feasible to give it another go at resolving this bug?

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7099584)

**A:** Since we re-opened this bug after the 1.3 patch seemingly didn't fully fix it, we are indeed actively investigating this right now. Thanks for bringing it back to our attention!

### Q: In my opinion there are three major areas of basic functionality that need to be addressed ... The first is having manually unwire all event handlers specified in the XAML ... The second issue is not having a proper transparent background for owner drawn windows with a proper pass thru when clicking on the transparent portion of the window ... The third major area is the complete lack of any helpers on the Window and AppWindow classes for some of the basic functionality that's existed for years.

Question link [here](https://github.com/microsoft/microsoft-ui-xaml/discussions/8884#discussioncomment-7117633)

**A:** We'll address each of these points separately here:

1. We are looking at this in the 1.5 time frame along with the overall shutdown experience.
2. Window transparency is not supported currently nor is it planned for 1.5. Click-through transparency is not on the roadmap since support for this in Windows is limited. You can use [SetWindowRgn](https://learn.microsoft.com/windows/win32/api/winuser/nf-winuser-setwindowrgn) like you can on any window, but you have to construct the correct HRGN for that and update it any time the region changes.
3. In 1.4 we provided an easy way to access AppWindow directly so you can access their helpers - no need to wrap wrappers. However, if you're talking about actual AppWindow limitations here, such as lacking APIs to support setting a minimum or maximum window size, then yes we understand. We plan on soliciting feedback on AppWindow functionality needs soon here on GitHub, so keep an eye out for that.

### Q: ​I think the issue is more about, only having 4 calls in a year, whilst having more content to cover - things will be glossed over, and not enough feedback call and response

**A:** The quarterly cadence and focus on the wider Windows dev platform is a new format for us, so we'll be evaluating how things go and make adjustments if it feels like it isn't enough.

### Q: How are airspace issues handled?

**A:** Island hosting is done via an HWND and therefore comes with all of the standard limitations and airspace issues of HWND hosting.

### Q: ​Also what about xamlislands for wpf/winforms? So far it sounds like you only got C++ for now

**A:** Per our [release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#xaml-islands-no-longer-experimental) for 1.4, at the moment we have only tested Islands in C++. We believe Islands should work in WPF and Winforms but verifying and providing convenient wrappers isn't a current focus area for us.

### Q: It does feel like its going to become more of a group of summarised blog posts, just "marketing" new stuff, and less about responding to feedback issues, and discussions

**A:** We do intend to post a GitHub discussion for people to ask questions before the calls going forward, as we have done with this call and previous ones, and will do our best to answer them.

### Q: ​Still no support for webauthenticator? Or does that fall under improving Maui support?

**A:** We don't have plans for this currently, no.

### Q: Are there any examples of the new widget interfaces added in WASDK 1.4 such as IWidgetProvider2?

**A:** Besides the original [release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#widgets-updates) that described what the new interfaces are and their basic functionality, we also have `IWidgetProvider2` documented [here](https://learn.microsoft.com/windows/apps/develop/widgets/implement-widget-provider-cs#implement-iwidgetprovider2) for an example that shows customization actions. Stay tuned to the [Windows Widgets Sample](https://github.com/microsoft/windowsappsdk-samples/tree/main/Samples/Widgets) for future updates demoing the new interfaces!

### Q: ​Making the Tab controls more robust is very welcome. I hope thought will be given to supporting Bottom and Left/Right side placement, to make it more flexible for people to use

**A:** Great suggestion! We don't have plans for this yet but we like the idea. However, we believe you are thinking of in-app tabs such as in Notepad, whereas we're referring to windows tabs. Just to clarify.

### Q: will the performance and feel of ScrollView be improved in the future?

**A:** We are making incremental improvements to scrolling performance, yes.

### Q: ​Drag & Drop support for WebView2 - what does it have to do with WinUI 3? Isn't that just improving WebView2 so that the new Outlook can support proper drag & drop?

**A:** Because of the windowless hosting of WebView2, there is work needed in WinUI 3 to implement drag & drop interfaces and pass the necessary events and info to WebView2.

### Q: ​Will we be able to build from the source or only view due to private dependencies?

**A:** To be clear, we are doing shared source, not open source, as there are underlying interwoven components linking WinUI 3 to the system that we can't share. We will not support building, deploying, or running the source code and will not accept PRs for it. We will support viewing, searching, and debugging of the source code when it is available. 

### Q: ​Do Microsoft internal teams get priority with feature and control requests, or can community sentiment push things up the pecking order?

**A:** We partially mentioned this on the call, but one of the big changes we made was putting internal requests, partner requests, and community requests in the same list. There's now a higher chance of prioritizing community sentiment than before we made this change.

### Q: ​GitHub has no issue priority, so HOW are you measuring that?

**A:** We mirror GitHub issues to our internal bug tracking system, which allows us to set severity and priority (although this information is only visible internally).

### Q: Will WinUI 3 gets similar API for setting title bar that goes around textbox or interactable objects without manually calculating bounds ourselves?

**A:** We do understand the request here to specify a composition visual for the titlebar, similar to what is done for UWP, but we don't have plans for that specifically in 1.5. That being said, we are looking to make the titlebar control easier to use. In 1.4 we merged our custom titlebars with the AppWindow titlebar in a step towards simplifying the situation (see the [1.4 release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#custom-titlebar--appwindow-titlebar-merger) for more details). Our goal was to achieve functionality parity with the older titlebar implementation, and in the future we plan on more feature improvements to make titlebar use more intuitive. In the XAML repo, please use the `area-TitleBar` tag on any suggestions you have for us to filter and prioritize them better, thank you!

### Q: ​Is there any chance to have a "draggable" content dialog?

**A:** We don't have plans for this currently.

**Comment by knightmeister** [comment]
Thanks for taking the time to write this up and publish it - the transparency is good.

One additional question: Will the team be taking community PRs for fixes when the source is made available? My time is limited but I am passionate about this so would like to contribute little pieces where I can. 

I expect a number of other contributors in the repository are in a similar position, it would be great for us to get through some of the lower priority issues from the Microsoft perspective.

Cheers 🙂

**Reply by ghord** [reply]
> We will not support building

I'm worried that this means you will not be able to build it yourself? How would we contribute if we cannot even test our code? Hope they only mean they will not support building for production.

**Reply by kmgallahan** [reply]
> Will the team be taking community PRs for fixes when the source is made available?

> A: To be clear, we are doing shared source, not open source, as there are underlying interwoven components linking WinUI 3 to the system that we can't share. We will not support building, deploying, or running the source code and will not accept PRs for it. We will support viewing, searching, and debugging of the source code when it is available.

**Reply by duncanmacmichael** [reply]
Sure thing, glad it was helpful! As far as I'm aware, at this time we won't be accepting PRs for the source (as kmgallahan mentioned with his excerpt). Although, I really do appreciate your helpful attitude here and will raise this to see if that might change in the future. Cheers 😊

**Reply by duncanmacmichael** [reply]
Correct, we won't support building the source for production at this time. The main reason is that WinUI 3 is undocked from the system and has underlying scaffolding that is needed to build it, but we can't share those components since they're tied into the OS. Our primary reason for sharing it is for education and debugging.

**Reply by duncanmacmichael** [reply]
@knightmeister Following up, here's what I gathered from the team regarding PRs for the source. Without being able to build the source, it won't be possible for any PRs to be built or tested before they are submitted. Given that, we won't be accepting PRs in general. But, realistically, if someone posts a PR which is unambiguously correct, small, and safe, we'll likely consider it. Hope this helps!

**Comment by Gaoyifei1011** [comment]
cross-process islands are mentioned in the 1.5 roadmap, we are curious about what this is, can you share a little bit about it

**Reply by duncanmacmichael** [reply]
Certainly - @codendone would you be able to shed some light here?

**Reply by codendone** [reply]
Cross-process islands are to support an island running in process A hosted in process B. This can provide security and reliability wins, similar to how WebView2 content runs in a separate process which doesn't have access to data in the host process and doesn't crash the host process if the WebView2 process crashes.

**Reply by duncanmacmichael** [reply]
Thanks @codendone!

**Reply by Gaoyifei1011** [reply]
thanks for sharing about this.

**Comment by Gaoyifei1011** [comment]
Can the future roadmap provide a winui design for custom title bar window menus?

**Reply by duncanmacmichael** [reply]
I'll direct this question to @pratikone, if we are able to comment on it yet.

**Reply by Gaoyifei1011** [reply]
thanks for sharing about this.Is there any follow-up whereabouts of this problem? I remember I tried it in the preview version of winui2.8 before, and there was no follow-up whereabouts

**Reply by pratikone** [reply]
A more elaborate question would be better so that I can understand it better. You can use the issues tab/discussion tab for it so that we can discuss it at length.

**Comment by BreeceW** [comment]
> However, we believe you are thinking of in-app tabs such as in Notepad, whereas we're referring to windows tabs. Just to clarify.

Can you clarify what “windows tabs” is referring to? The roadmap describes “highly functional tabbed windows that work with touch, snapping, and other OS features.” I interpreted this as making a standardized, easy way to make tabbed apps like Notepad/File Explorer/Terminal with features like tab tear-off built-in. If this is not what that means, can it be made clear what this feature in 1.5 is expected to do?

**Reply by duncanmacmichael** [reply]
Sure, sorry if the wording was confusing or if we misinterpreted the original question. My understanding is that the 1.5 feature refers to window tabs, like you can have one window with a Word tab, an Edge tab, and a Visual Studio tab, and not to in-app tabs ala Notepad/File Explorer/Terminal. @ranjeshj can you help clarify?

**Comment by ReneeGA2020** [comment]
It concerns me that there is basically nothing for the windows UI in the .net conf 2023 agenda.

**Reply by kmgallahan** [reply]
WinUI is owned by the Windows part of Microsoft. There is no direct relationship between .NET and WinUI, besides C# being one of the supported languages.

They own MAUI though, which uses WinUI on Windows. WinForms and WPF were .NET products.

**Reply by duncanmacmichael** [reply]
Correct, thank you @kmgallahan!

**Reply by ReneeGA2020** [reply]
It's cool to know this infomation. Does winui have plan to support .NET 8? It's weird that MAUI already support .NET 8 while it is supposed be winui on windows.

**Reply by duncanmacmichael** [reply]
Yes, we do plan on supporting .NET 8 once it hits GA. We don't officially support it currently as of version 1.4 of the WASDK because it's still prerelease and there was a breaking change with the runtime identifier graph, but you can still try it out with the workaround we recommend in the [1.4 release notes](https://learn.microsoft.com/windows/apps/windows-app-sdk/stable-channel#known-issues).

Tagging @bpulliam to double-check me on this :)

**Comment by ghord** [comment]
> Drag & Drop support for WebView2

Does this mean the team will spend some resources on working on Drag&Drop in general? Currently Drag&Drop is broken for elevated apps, and from what I understand it would require winui to reimplement in-app Drag&Drop without relying on OS functionality. Is this something that is being done as a part of this changes, or will WebView2 Drag&Drop still be unsupported in elevated apps?

**Reply by duncanmacmichael** [reply]
Good question - @codendone would you be the right one to ask about this? I'm not aware of plans to address Drag & Drop in elevated apps but I could be wrong.

**Reply by codendone** [reply]
The planned Drag & Drop work is specifically in the WebView2 control.

Support for Drag & Drop in elevated apps is not in plan for 1.5.

**Comment by Gaoyifei1011** [comment]
At present, in winui3, the mouse of the modified control has a dependency property protectedcursor, the type of which is protected. If you need to modify it, you need to use reflection to modify it. Can you directly change the permission of this property to public in the future?
Could you please explain in detail why this dependency property is of protected type, I don't quite understand
https://github.com/microsoft/WindowsAppSDK/discussions/3910

**Reply by duncanmacmichael** [reply]
Hey, thanks for asking this! This discussion is meant to address questions either brought up on the community call or as a direct follow-up, so for additional questions I'd recommend opening an issue and labeling it as a question. Having said that, I'm not sure who would be the best person to ask about this (and the linked thread). @codendone do you know whom we can ask about cursors?

**Reply by codendone** [reply]
The goal is eventually add a public `UIElement.Cursor` property. This feature is on the backlog and we don't currently have a timeline for it.

**Reply by duncanmacmichael** [reply]
Thanks @codendone!

**Comment by Gaoyifei1011** [comment]
Is there a way to use the components of the Windows app sdk in the.NET framework (except for the winui3 section), for example I want to customize the title bar in the desktop application of the.NET framework using the     appwindow of the Windows App SDK

**Reply by duncanmacmichael** [reply]
I don't believe so, @bpulliam can probably confirm.

**Comment by sotanakamura** [comment]
Thanks for all your hard work. I have recently been working on a python language projection for WinUI 3. I hope the WinUI project will be more exciting in the future.

**Comment by Gaoyifei1011** [comment]
> I don't believe so, @bpulliam can probably confirm.

OK，Thanks for your answering.

There are a few other questions here.

When customizing the SystemBackdrop, switching themes will trigger the App.UnhandledException event, this bug has been dragging for several versions, I want to know when this bug can be fixed?
![image](https://github.com/microsoft/microsoft-ui-xaml/assets/49179966/17f819cf-de8b-42d2-87de-16c01cfeafce)

The default MicaBackdrop has a Kind attribute to customize the style of Mica, but DesktopAcrylicBackdrop lacks this attribute, can it be added in the next version?

Is it possible to add a default SytstemBackdropConfiguration property for the default MicaBackdrop and DesktopAcrylicBackdrop, allowing for custom systembackdrop?

**Reply by duncanmacmichael** [reply]
For the bug, if there's an open issue for it already then I'd recommend pinging for status there. The devs triage and watch those more closely than these Q&A comment sections. If there isn't one, then I'd recommend opening a new bug.

As for your other questions, I'd recommend opening a new issue and labeling it as a question, then we can tag one of the devs who works on materials for a more detailed answer. :)

**Reply by Gaoyifei1011** [reply]
This issue has been a problem for a long time and there has been no progress https://github.com/microsoft/microsoft-ui-xaml/issues/8415

**Comment by pubtom** [comment]
When will be the next Community Call?

**Reply by duncanmacmichael** [reply]
Coming soon, thanks for your patience! :) Sorry we've been quiet. We've been working on a more sustainable plan to engage with the developer community long term, and we're excited to share the plan with you all soon!

**Comment by ChaimCL** [comment]
Any shareable news about the next community call?

**Reply by JanRajnoha** [reply]
@duncanmacmichael Do you have any info?

**Reply by MEK3DK** [reply]
I tried to raise the issue here also: [New community call](https://github.com/microsoft/microsoft-ui-xaml/discussions/9181)

But there is absolute silence, so can only assume WinUI is no longer in development.

**Reply by knightmeister** [reply]
@MEK3DK sadly I'm inclined to agree with you. I raised an issue back in August which was turned into [a discussion](https://github.com/microsoft/microsoft-ui-xaml/discussions/8712).

The state of this repository and the communication from Microsoft is highly concerning. Fortunately my app is properly factored so I can move off WinUI relatively easy.

If this product isn't dead, management need to do their job and allay reasonable community concerns.

**Reply by duncanmacmichael** [reply]
As I mentioned above, sorry we've been quiet. We've been working on a more sustainable plan to engage with the developer community long term and have listened to feedback about the format. Will have more news very soon! 

WinUI is definitely still in very active development. ;)

**Reply by JanRajnoha** [reply]
Okay, understood, but could you give us one promise for next call? Only for next one cause it will be after long time.

Please, make it longer, at least 2-3 hours and answer as much as possible questions, because community need some info about project and need to see everything is okay... Especially when we are talking about long time waiting bugs, future, tools support and so on. I expect people will have ton of questions for you.

**Comment by pubtom** [comment]
Half year passed after the last WinUI Community call ...

**Reply by JanRajnoha** [reply]
And you probably missed recent - #9361
